File: 31-08-2025_Sudoku Solver.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int n = 3;
    int N = 9;
    int rows[9][10] = {0};
    int cols[9][10] = {0};
    int boxes[9][10] = {0};
    vector<vector<char>>* board;
    bool solved = false;

    bool couldPlace(int d, int row, int col) {
        int idx = (row / n) * n + (col / n);
        return rows[row][d] + cols[col][d] + boxes[idx][d] == 0;
    }

    void placeNumber(int d, int row, int col) {
        int idx = (row / n) * n + (col / n);
        rows[row][d]++;
        cols[col][d]++;
        boxes[idx][d]++;
        (*board)[row][col] = d + '0';
    }

    void removeNumber(int d, int row, int col) {
        int idx = (row / n) * n + (col / n);
        rows[row][d]--;
        cols[col][d]--;
        boxes[idx][d]--;
        (*board)[row][col] = '.';
    }

    void placeNextNumbers(int row, int col) {
        if (row == N - 1 && col == N - 1) {
            solved = true;
        } else if (col == N - 1) {
            backtrack(row + 1, 0);
        } else {
            backtrack(row, col + 1);
        }
    }

    void backtrack(int row, int col) {
        if ((*board)[row][col] == '.') {
            for (int d = 1; d <= 9; d++) {
                if (couldPlace(d, row, col)) {
                    placeNumber(d, row, col);
                    placeNextNumbers(row, col);
                    if (!solved) removeNumber(d, row, col);
                }
            }
        } else {
            placeNextNumbers(row, col);
        }
    }

    void solveSudoku(vector<vector<char>>& b) {
        board = &b;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (b[i][j] != '.') {
                    int d = b[i][j] - '0';
                    placeNumber(d, i, j);
                }
            }
        }
        backtrack(0, 0);
    }
};

File: 31-08-2025_Sudoku Solver.py
class Solution(object):
    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: None Do not return anything, modify board in-place instead.
        """
        # 9x9 arrays to track digits usage
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]
        empty = []

        # initialize tracking structures
        for r in range(9):
            for c in range(9):
                if board[r][c] == ".":
                    empty.append((r, c))
                else:
                    d = board[r][c]
                    rows[r].add(d)
                    cols[c].add(d)
                    boxes[(r // 3) * 3 + (c // 3)].add(d)

        def backtrack(idx=0):
            if idx == len(empty):
                return True  # all filled

            r, c = empty[idx]
            b = (r // 3) * 3 + (c // 3)

            for d in "123456789":
                if d not in rows[r] and d not in cols[c] and d not in boxes[b]:
                    # place number
                    board[r][c] = d
                    rows[r].add(d)
                    cols[c].add(d)
                    boxes[b].add(d)

                    if backtrack(idx + 1):
                        return True

                    # undo placement
                    board[r][c] = "."
                    rows[r].remove(d)
                    cols[c].remove(d)
                    boxes[b].remove(d)

            return False

        backtrack()
